### PR TITLE
feat: add Dependency Model

### PR DESCRIPTION
**Pull request by Kravcu**
Hello everyone,

I'd like to contribute and help with adding dependencies, I'd appreciate if you could take a look at this code and say what's wrong here.

### CODE REVIEW

**Comment by madpah** [comment]
Hi @Kravcu - thanks for the contribution. Can you open an Issue Ticket to describe the feature/fix etc. that you are contributing please? It will help us (and the community) understand and track the change.

**Comment by madpah** [comment]
I think perhaps you are looking to implement #7 @Kravcu - awesome!

Will review the PR and provide any comments.

File: cyclonedx/model/component.py [hunk 1]
```diff
@@ -126,6 +128,7 @@ def __init__(self, name: str, component_type: ComponentType = ComponentType.LIBR
         self.purl = purl
         self.external_references = external_references if external_references else []
         self.properties = properties
+        self.__dependencies = dependencies or []
```
Comment (madpah) [inline cyclonedx/model/component.py:131]: I would move this to be with line 153 (suggestion).

File: cyclonedx/model/component.py [hunk 2]
```diff
@@ -477,6 +480,37 @@ def has_vulnerabilities(self) -> bool:
...
+        Does this Component have any dependencies?
+
+        Returns:
+             `True` if this Component has 1 or more dependency, `False` otherwise.
+        """
+        return bool(self.__dependencies)
+
+    def get_dependencies(self) -> List[Dependency]:
+        """
+        Get all the Dependencies for this Component.
```
Comment (madpah) [inline cyclonedx/model/component.py:494]: We've been trying to utilise the descriptions for attributes as provided in the CycloneDX Specification - this doc would be:

> Defines the direct dependencies of a component. Components that do not have their own dependencies MUST be declared as empty elements within the graph. Components that are not represented in the dependency graph MAY have unknown dependencies. It is RECOMMENDED that implementations assume this to be opaque and not an indicator of a component being dependency-free.

File: cyclonedx/model/dependency.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) OWASP Foundation. All Rights Reserved.
+
+# See https://github.com/package-url/packageurl-python/issues/65
+from packageurl import PackageURL  # type: ignore
+
+
+class Dependency:
```
Comment (madpah) [inline cyclonedx/model/dependency.py:24]: Missing class documentation.

File: cyclonedx/model/dependency.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    def __init__(self, purl: PackageURL = None) -> None:
+        self.purl = purl
```
Comment (madpah) [inline cyclonedx/model/dependency.py:26]: Can you please add @property getter/setting methods for `purl`.

File: cyclonedx/output/json.py [hunk 5]
```diff
@@ -56,16 +56,22 @@ def generate(self, force_regeneration: bool = False) -> None:
...
                 )
+        dependencies: Dict[str, List[Dict[Any, Any]]] = {"dependencies": []}
+        for component in self.get_bom().components:
+            if component.has_dependencies() and component.purl:
+                dependencies["dependencies"].append(CycloneDxJSONEncoder.serialize_component_dependencies(component))
 
         bom_json = json.loads(json.dumps(self.get_bom(), cls=CycloneDxJSONEncoder))
         bom_json = json.loads(self._specialise_output_for_schema_version(bom_json=bom_json))
+
+        data_to_generate_json = {**self._create_bom_element(), **bom_json}
```
Comment (madpah) [inline cyclonedx/output/json.py:67]: Love this change 👍

File: cyclonedx/output/serializer/json.py [hunk 6]
```diff
@@ -59,6 +59,10 @@ def default(self, o: Any) -> Any:
         if isinstance(o, XsUri):
             return str(o)
 
+        # PackageURL
```
Comment (madpah) [inline cyclonedx/output/serializer/json.py:62]: Interesting you've added this code @Kravcu - myself and the other maintainers struggled with JSON serialisation of `PackageURL` objects as they are extensions of `tuple`, which the JSON serializer handles before the `default()` method is called - so this code will never be reached.

Our workaround was lines 86-89 below.

File: cyclonedx/output/serializer/json.py [hunk 7]
```diff
@@ -89,3 +93,10 @@ def default(self, o: Any) -> Any:
 
         # Fallback to default
         super().default(o=o)
+
+    @classmethod
+    def serialize_component_dependencies(cls, c: Component) -> Dict[str, Union[str, List[str]]]:
```
Comment (madpah) [inline cyclonedx/output/serializer/json.py:98]: The issue here is that `bom-ref`'s for Components are not guaranteed to the PURLs.

Each Component now has a `bom-ref` attribute which should be used.

**Review by madpah — CHANGES_REQUESTED** [review]

**Comment by madpah** [comment]
@Kravcu - I've added some review points to the PR - the main point is the assumption that Components will be referenced using a PURL, which is not the case I'm afraid.

Aside from the PR comments, there are a couple of other hygiene points to look at:
1. You need to sign your commits - please check [CONTRIBUTING](https://github.com/CycloneDX/cyclonedx-python-lib/blob/master/CONTRIBUTING.md)
2. Few static analysis violations have been reported

Thanks again and don't hesitate to ask any questions.

**Comment by jkowalleck** [comment]
@madpah  @Kravcu 
we might need to discuss and document, if a set of dependencies should modify the hash of a component,
and therefore be part of `__hash__`.

**Comment by madpah** [comment]
This has been implemented now through #210.

**Pull request closed without merging**